cypress/integration/ete/register_email_sent.spec.js

```diff
@@ -29,28 +34,36 @@ describe('Registration email sent page', () => {
     });
     cy.visit(`/register/email-sent`);
     cy.contains('Email sent');
-    cy.contains(email.value);
+    cy.contains(existing.email);
     cy.contains('Resend email');
     cy.contains('Change email address');
   });
 
   // Depends on a Guest account already created using this email.
   it('should resend the email when the resend button is clicked', () => {
-    cy.emptyInbox(email.inbox).then(() => {
-      cy.setCookie('PLAY_SESSION_2', encryptedEmail, {
-        log: true,
-      });
-      cy.visit(`/register/email-sent`);
-      cy.contains(email.value);
-      cy.contains('Resend email').click();
-      cy.waitForLatestEmail(email.inbox).then((response) => {
-        expect(response.body).to.have.string('Complete registration');
-        // Extract the welcome token, so we can redirect to the welcome flow.
-        const match = response.body.match(/theguardian.com\/welcome\/([^"]*)/);
-        const token = match[1];
-        cy.visit(`/welcome/${token}`);
-        cy.contains('Create password');
-      });
+    cy.setCookie('PLAY_SESSION_2', encryptedEmail, {
+      log: true,
+    });
+    cy.visit(`/register/email-sent`);
+    cy.contains(existing.email);
+    cy.contains('Resend email').click();
+    cy.mailosaurGetMessage(
+      existing.serverId,
+      {
+        sentTo: existing.email,
+      },
+      {
+        receivedAfter: new Date(),
```

Comment:
haha it's a bit unintuitive, but when I was writing it it made sense to me. Here's the timeline:

We programmatically click the Resend Email button on line 49 and a network request is fired off to IDAPI to send our email (this will take milliseconds to seconds to deliver cause networks are slow)

... nanoseconds later ...

Line 56 (`new Date()`) is executed , so an _extremely_ small amount of time will have passed before `new Date()` is evaluated after the Resend Email button is clicked.

`now` is now set to a few nanoseconds after the initial click of the button.

.... shortly thereafter ... 

We construct our query for Mailosaur to query its mail list for anything after the `now` we just evaluated and fire it off as an async request.

... our query to Mailosaur will wait a few milliseconds to seconds depending on whether the email has been delivered yet... (max timeout is 10s iirc)

Reply:
To my mind, there are two problems here:

1. The code is confusing. It doesn't make sense to see something saying only return emails `receivedAfter: now`. If you created the timestamp at an earlier point you could have `receivedAfter: theRequestWasMade` which I think is a lot clearer
2. It might be true that the code works, but it only works by chance, you're getting away with it basically. I think it's better to write code that works for logical, structured reasons. You want to _know_ that your date filter is correct in the absolute sense, not just good enough.

The conditions that this code executes in today might mean that it succeeds but the future can bring lots of changes so you want to build in as much resilience as possible. This might seem like overkill, isn't this just a test so why should we care? But we should care because in my experience it's surprising how often these sorts of things turn around and bite you - a lot more than you'd think!

Reply:
Great suggestion to move it into its own variable and set it earlier, nice readability improvement there and we'll avoid any possibility of failure.